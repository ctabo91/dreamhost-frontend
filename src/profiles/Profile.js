import React, { useState, useContext } from "react";
import Alert from "../common/Alert";
import DreamHostApi from "../api/api";
import UserContext from "../auth/UserContext";

/** Shows user profile info, and allows the user to update the info as desired. 
 * 
 * If there are any errors submitting the form, they will be displayed directly above the submit button.
 */ 

function Profile() {
    const { currentUser, setCurrentUser } = useContext(UserContext);
    const [formData, setFormData] = useState({
        firstName: currentUser.firstName,
        lastName: currentUser.lastName,
        email: currentUser.email,
        username: currentUser.username,
        password: "",
    });
    const [formErrors, setFormErrors] = useState([]);

    const [saveConfirmed, setSaveConfirmed] = useState(false);

    const handleSubmit = async (e) => {
        e.preventDefault();

        let profileData = {
            firstName: formData.firstName,
            lastName: formData.lastName,
            email: formData.email,
            password: formData.password,
        };

        let username = formData.username;
        let updatedUser;

        try {
            updatedUser = await DreamHostApi.updateUser(username, profileData);
        } catch (errors) {
            setFormErrors(errors);
            return;
        }

        setFormData(formData => ({
            ...formData,
            password: "",
        }));
        setFormErrors([]);
        setSaveConfirmed(true);

        setCurrentUser(updatedUser);
    };

    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value,
        }));
        setFormErrors([]);
    };

    return (
        <div className="form-container col-md-6 offset-md-3 col-lg-6 offset-lg-3 mt-5 pt-5">
            <h4 className="form-title display-4 text-center mb-3">- Profile -</h4>
            <div className="form-card card">
                <div className="card-body">
                    <form>
                        <div className="form-group">
                            <p id="username" className="form-control-plaintext text-center">Username: {formData.username}</p>
                        </div>
                        <div className="form-group">
                            <label htmlFor="firstName">First Name</label>
                            <input 
                                id="firstName"
                                name="firstName"
                                className="form-control"
                                value={formData.firstName}
                                onChange={handleChange}
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="lastName">Last Name</label>
                            <input 
                                id="lastName"
                                name="lastName"
                                className="form-control"
                                value={formData.lastName}
                                onChange={handleChange}
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="email">Email</label>
                            <input 
                                id="email"
                                name="email"
                                className="form-control"
                                value={formData.email}
                                onChange={handleChange}
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="password">Password</label>
                            <input
                                id="password" 
                                type="password"
                                name="password"
                                className="form-control"
                                value={formData.password}
                                onChange={handleChange}
                            />
                        </div>

                        {formErrors.length
                            ? <Alert type="danger" messages={formErrors} />
                            : null}

                        {saveConfirmed
                            ? <Alert type="success" messages={["Updated successfully."]} />
                            :null}

                        <button
                            className="submit-button btn btn-primary btn-block mt-4"
                            onClick={handleSubmit}
                        >
                            Save Changes
                        </button>
                    </form>
                </div>
            </div>
        </div>
    );
}

export default Profile;