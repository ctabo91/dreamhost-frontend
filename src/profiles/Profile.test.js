import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import Profile from './Profile';
import DreamHostApi from '../api/api';
import UserContext from '../auth/UserContext';

jest.mock('../api/api');

const mockCurrentUser = {
  username: 'testUser',
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@example.com',
};

const mockSetCurrentUser = jest.fn();

const renderProfileComponent = () => {
  return render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: mockCurrentUser, setCurrentUser: mockSetCurrentUser }}>
        <Profile />
      </UserContext.Provider>
    </MemoryRouter>
  );
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Profile Component', () => {
  it('matches snapshot', async () => {
    const asFragment = await act(async () => {
      const { asFragment } = renderProfileComponent();
      return asFragment;
    })

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders user profile information correctly', async () => {
    await act(async () => renderProfileComponent());

    expect(screen.getByText('Username: testUser')).toBeInTheDocument();
    expect(screen.getByLabelText('First Name')).toHaveValue('John');
    expect(screen.getByLabelText('Last Name')).toHaveValue('Doe');
    expect(screen.getByLabelText('Email')).toHaveValue('john.doe@example.com');
    expect(screen.getByLabelText('Password')).toHaveValue('');
  });

  it('updates form data on input change', async () => {
    await act(async () => renderProfileComponent());

    await act(async () => {
      userEvent.type(screen.getByLabelText('First Name'), 'NewFirstName');
      userEvent.type(screen.getByLabelText('Last Name'), 'NewLastName');
      userEvent.type(screen.getByLabelText('Email'), 'new.email@example.com');
      userEvent.type(screen.getByLabelText('Password'), 'newPassword');
    });

    expect(screen.getByLabelText('First Name')).toHaveValue('JohnNewFirstName');
    expect(screen.getByLabelText('Last Name')).toHaveValue('DoeNewLastName');
    expect(screen.getByLabelText('Email')).toHaveValue('john.doe@example.comnew.email@example.com');
    expect(screen.getByLabelText('Password')).toHaveValue('newPassword');
  });

  it('handles form submission successfully', async () => {
    DreamHostApi.updateUser.mockResolvedValueOnce({
      ...mockCurrentUser,
      firstName: 'NewJohn',
      lastName: 'NewDoe',
      email: 'new.john.doe@example.com',
    });

    await act(async () => renderProfileComponent());

    await act(async () => {
      userEvent.clear(screen.getByLabelText("First Name"));
      userEvent.clear(screen.getByLabelText("Last Name"));
      userEvent.clear(screen.getByLabelText("Email"));
    });

    await act(async () => {
      userEvent.type(screen.getByLabelText('First Name'), 'NewJohn');
      userEvent.type(screen.getByLabelText('Last Name'), 'NewDoe');
      userEvent.type(screen.getByLabelText('Email'), 'new.john.doe@example.com');
    });

    await act(async () => {
      userEvent.click(screen.getByText('Save Changes'));
    });

    expect(DreamHostApi.updateUser).toHaveBeenCalledWith('testUser', {
        firstName: 'NewJohn',
        lastName: 'NewDoe',
        email: 'new.john.doe@example.com',
        password: '',
    });

    expect(screen.getByLabelText('First Name')).toHaveValue('NewJohn');
    expect(screen.getByLabelText('Last Name')).toHaveValue('NewDoe');
    expect(screen.getByLabelText('Email')).toHaveValue('new.john.doe@example.com');
    expect(screen.getByLabelText('Password')).toHaveValue('');
    expect(mockSetCurrentUser).toHaveBeenCalledWith({
      ...mockCurrentUser,
      firstName: 'NewJohn',
      lastName: 'NewDoe',
      email: 'new.john.doe@example.com',
    });
  });

  it('handles form submission with errors', async () => {
    DreamHostApi.updateUser.mockRejectedValueOnce(['Error 1', 'Error 2']);

    await act(async () => renderProfileComponent());

    await act(async () => {
      userEvent.clear(screen.getByLabelText("First Name"));
      userEvent.clear(screen.getByLabelText("Last Name"));
      userEvent.clear(screen.getByLabelText("Email"));
    });

    await act(async () => {
      userEvent.type(screen.getByLabelText('First Name'), 'NewJohn');
      userEvent.type(screen.getByLabelText('Last Name'), 'NewDoe');
      userEvent.type(screen.getByLabelText('Email'), 'new.john.doe@example.com');
    });

    await act(async () => {
      userEvent.click(screen.getByText('Save Changes'));
    });

    expect(DreamHostApi.updateUser).toHaveBeenCalledWith('testUser', {
      firstName: 'NewJohn',
      lastName: 'NewDoe',
      email: 'new.john.doe@example.com',
      password: '',
    });

    expect(screen.getByLabelText('First Name')).toHaveValue('NewJohn');
    expect(screen.getByLabelText('Last Name')).toHaveValue('NewDoe');
    expect(screen.getByLabelText('Email')).toHaveValue('new.john.doe@example.com');
    expect(screen.getByLabelText('Password')).toHaveValue('');
    expect(screen.getByText('Error 1')).toBeInTheDocument();
    expect(screen.getByText('Error 2')).toBeInTheDocument();
    expect(mockSetCurrentUser).not.toHaveBeenCalled();
  });
});
