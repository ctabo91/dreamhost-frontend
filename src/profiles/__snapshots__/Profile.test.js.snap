// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-container col-md-6 offset-md-3 col-lg-6 offset-lg-3 mt-5 pt-5"
  >
    <h4
      class="form-title display-4 text-center mb-3"
    >
      - Profile -
    </h4>
    <div
      class="form-card card"
    >
      <div
        class="card-body"
      >
        <form>
          <div
            class="form-group"
          >
            <p
              class="form-control-plaintext text-center"
              id="username"
            >
              Username: testUser
            </p>
          </div>
          <div
            class="form-group"
          >
            <label
              for="firstName"
            >
              First Name
            </label>
            <input
              class="form-control"
              id="firstName"
              name="firstName"
              value="John"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              class="form-control"
              id="lastName"
              name="lastName"
              value="Doe"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              value="john.doe@example.com"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <button
            class="submit-button btn btn-primary btn-block mt-4"
          >
            Save Changes
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
